eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala

```diff
@@ -512,7 +512,8 @@ class Channel(val nodeParams: NodeParams, val wallet: OnChainChannelFunder, val
           }
           if (d.remoteShutdown.isDefined && !commitments1.localHasUnsignedOutgoingHtlcs) {
             // we were waiting for our pending htlcs to be signed before replying with our local shutdown
-            val localShutdown = Shutdown(d.channelId, commitments1.localParams.defaultFinalScriptPubKey)
+            val finalScriptPubKey = if (commitments1.channelFeatures.hasFeature(Features.UpfrontShutdownScript)) commitments1.localParams.defaultFinalScriptPubKey else nodeParams.currentFinalScriptPubKey
```

Comment:
You should still make `LocalParams.finalScriptPubKey` an option defined only when `Feature.UpfrontShutdownScript` is present. It can be done in a backward-compatible way, codecs wise, by using `length=0` for an empty option.

This has the following advantages:
- remove all ambiguity: when we have no `finalScriptPubkey` available, we have to generate one
- force us to evaluate right now whether we _can_ drop all existing `LocalParams.finalScriptPubKey` and forget about it

Reply:
done in https://github.com/ACINQ/eclair/pull/2565/commits/1f1b68dc22d32dbe0a0a6de4a091096711ae5f24